### ISSUE TITLE
Upgrade: new instance of sozu doesn't listen (master / workers)

### ISSUE THREAD

**Issue by BlackYoup** [sozu-proxy/sozu#394]
When I upgrade (to a new built version of sozu, not just upgrading the same version), sozu doesn't start well.

The master doesn't seem to have its communication channel opened. Workers are starting but I guess they don't receive their configuration and they don't listen.

The upgrade process goes well.

Here are the logs (`DEBUG` level):

sozu logs
```
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549871808925 17936 MASTER DEBUG sozu::command   server accepting socket
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549890310794 17936 MASTER INFO  launching new master
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549891825427 17936 MASTER INFO  master launched: 12443
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549965927902 12443 MASTER DEBUG sozu::command::orders   listener is: 9
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549965974742 12443 MASTER DEBUG sozu::command::orders   registering: Ok(())
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549966237509 12443 MASTER ERROR could not change blocking status for stream: Error { repr: Os { code: 9, message: "Bad file descriptor" } }
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549966427737 12443 MASTER INFO  starting new master loop
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549966541725 17936 MASTER DEBUG sozu::command::orders   upgrade channel sent: Some(true)
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549966726843 12443 MASTER DEBUG sozu::command   server accepting socket
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549966801602 17936 MASTER INFO  wrote final message, closing
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549969765428 12443 MASTER INFO  1 launching worker
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549969788948 12443 MASTER DEBUG sozu::worker    executable path is /usr/x86_64-pc-linux-gnu/bin/sozu
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549970150554 12443 MASTER INFO  1 worker launched: 12444
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 12 send empty
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549970561539 12443 MASTER INFO  sent listeners from master: Ok(())
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549970598499 12443 MASTER INFO  created new worker: 1
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549970618389 12443 MASTER DEBUG sozu::command::orders   registering new sock UnixStream { inner: UnixStream { fd: 10, local: (unnamed), peer: (unnamed) } } at token 3 for tag BLAH and id 1 (sock error: Ok(None))
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549970842096 12443 MASTER DEBUG sozu::command::orders   proxyconfig client order SoftStop
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549970867716 12443 MASTER ERROR state cannot handle order message: SoftStop
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549970988725 17937 WRK-00 ERROR state cannot handle order message: SoftStop
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549971055094 17937 WRK-00 INFO  SOFT-STOP-WORKER-FOndAD processing soft shutdown
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549971084193 17937 WRK-00 INFO  SOFT-STOP-WORKER-FOndAD processing soft shutdown
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549971101373 17937 WRK-00 INFO  SOFT-STOP-WORKER-FOndAD processing soft shutdown
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549971147563 17937 WRK-00 INFO  last client stopped, shutting down!
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549971179642 17937 WRK-00 INFO  0 ending event loop
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549988279277 12444 WRK-01 INFO  worker 1 starting...
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: 2018-03-12T12:52:21Z 5769549988346527 12444 WRK-01 DEBUG sozu::worker    setting up metrics: local address = Ok(
Mar 12 13:52:21 exherbo-dedicated sozu[17936]:     V4(
Mar 12 13:52:21 exherbo-dedicated sozu[17936]:         0.0.0.0:43713
Mar 12 13:52:21 exherbo-dedicated sozu[17936]:     )
Mar 12 13:52:21 exherbo-dedicated sozu[17936]: )
```

sozuctl upgrade:
```
~ # sozuctl -c /etc/sozu/config.toml upgrade
Preparing to upgrade proxy...

+--------+-------+-----------+
| Worker | pid   | run state |
+--------+-------+-----------+
| 0      | 17937 | Running   |
+--------+-------+-----------+

Upgrading master process
Master process upgrade succeeded: new master process launched with pid 12443, closing the old one
Reconnecting to new master process...
Upgrading worker 1 (of 1)
Proxy successfully upgraded!
```

After the upgrade, I can't connect to sozu using sozuctl:
```
~ # sozuctl -c /etc/sozu/config.toml status
thread 'main' panicked at 'could not connect to the command unix socket: Error { repr: Os { code: 111, message: "Connection refused" } }', src/libcore/result.rs:916:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by Geal** [comment]
I can't reproduce here. Is the unix socket file still present?

**Comment by BlackYoup** [comment]
IIRC yes, it's still present.

Do you want me to add some logs in some places to track what might going wrong?

**Comment by Geal** [comment]
Yes please

**Comment by BlackYoup** [comment]
Do you have any idea as to where I should put logs?

**Comment by Geal** [comment]
probably in the master's event loop, and around its channel registration

**Comment by BlackYoup** [comment]
I can easily reproduce the problem when sozu gets started with systemd:
- start sozu with systemd
- launch an upgrade
- sozu doesn't listen anymore because it somehow exited: I can't find any sozu process running after the upgrade happened

systemctl status sozu:
```
# systemctl status sozu
● sozu.service - Sozu - A HTTP reverse proxy, configurable at runtime, fast and safe, built in Rust.
   Loaded: loaded (/usr/x86_64-pc-linux-gnu/lib/systemd/system/sozu.service; disabled; vendor preset: enabled)
   Active: inactive (dead)
     Docs: https://docs.rs/sozu/

Mar 27 14:03:08 exherswag sozu[4205]:         Backend {
Mar 27 14:03:08 exherswag sozu[4205]:             app_id: "MyApp",
Mar 27 14:03:08 exherswag sozu[4205]:             backend_id: "MyApp-0",
Mar 27 14:03:08 exherswag sozu[4205]:             ip_address: "127.0.0.1",
Mar 27 14:03:08 exherswag sozu[4205]:             port: 1026
Mar 27 14:03:08 exherswag sozu[4205]:         }
Mar 27 14:03:08 exherswag sozu[4205]:     )
Mar 27 14:03:08 exherswag sozu[4205]: }
Mar 27 14:03:08 exherswag sozu[4205]: 2018-03-27T12:03:08Z 120140278840250 5412 WRK-01 DEBUG        sozu_lib::network::http        INIT-2 add backend Backend { app_id: "MyApp", backend_id: "MyApp-0", ip_address: "127.0.0.1", port: 1026 }
Mar 27 14:03:08 exherswag sozu[4205]: 2018-03-27T12:03:08Z 120140278862896 5412 WRK-01 INFO        starting event loop
```

Maybe a bad systemd configuration?

**Comment by BlackYoup** [comment]
@Keruspe do you have some insights about this issue?

**Comment by BlackYoup** [comment]
I checked a bit more and what seems to happen is that systemd, after the old master process exited, fails to find the new main process and just stop the service seeing the old master process is now gone. (This is what I suppose, I couldn't find a lot of resources over the internet)

To help with that, we could use the systemd `PIDFile` directive that allows a program to store its main process's PID file. The idea is that, when the new master spawns (or before the old one exits?), it writes its pid into this file, usually `/run/sozu/sozu.pid`.
When the old master process exits, systemd will read the pid file and set the new master process accordingly.
Systemd will also clean this file if needed when the service is stopped.

I made a POC here https://github.com/BlackYoup/sozu/commit/716ced47959dabb6b6a8ba143f8dadd3c59da029 and the upgrade process goes well, the new processes aren't killed and sozu responds correctly to requests.

If we go this way, there are multiple questions about this:
- when do we write the pid of the new master process? When the new master starts? When the old master is about to stop itself? What if there is an error during the new master startup?
- What happens if we fail to create / write that file?
- There should be a configuration option to set that file or disable the feature.
- What about development? We most likely don't want to create a pid file. See the previous line
